src/storage-controller/src/statistics.rs

```diff
@@ -28,37 +28,13 @@ use tokio::sync::oneshot;
 
 use crate::collection_mgmt::CollectionManager;
 
-/// An enum that tracks the lifecycle of statistics objects
-/// in the controller.
-///
-/// When sources/sinks are dropped, some state (including statistics) are
-/// cleaned up later, as part of processing the storage controller. This
-/// means that a new statistics update could happen between `DROP SOURCE`
-/// and cleanup, which is we need to distinguished `Uninitialized` from
-/// "explicitly removed" so that we no longer consolidate stats for an
-/// item into the `shared_stats` shared map.
-#[derive(Debug)]
-pub(super) struct StatsInitState<T>(pub BTreeMap<usize, T>);
-
-impl<T> StatsInitState<T> {
-    /// Set the value for the given id, overriding it if it already exists,
-    /// and doing nothing if its been removed.
-    pub(super) fn set_if_not_removed(this: Option<&mut Self>, worker_id: usize, val: T) {
-        match this {
-            Some(StatsInitState(map)) => {
-                map.insert(worker_id, val);
-            }
-            None => {}
-        }
-    }
-}
-
 /// Spawns a task that continually (at an interval) writes statistics from storaged's
 /// that are consolidated in shared memory in the controller.
 pub(super) fn spawn_statistics_scraper<Stats, T>(
     statistics_collection_id: GlobalId,
     collection_mgmt: CollectionManager<T>,
-    shared_stats: Arc<Mutex<BTreeMap<GlobalId, StatsInitState<Stats>>>>,
+    shared_stats: Arc<Mutex<BTreeMap<GlobalId, Option<Stats>>>>,
+    previous_values: Vec<Row>,
```

Comment:
This can fail to will reconcile correctly because these previous values are used in a blind `append_to_collection` below which does not assert that the current upper of the shard is the upper that was observed during the snapshot of the previous values. This means you might append the wrong update because the upper moved from underneath you.

To fix this the scraper needs to be able to obtain a snapshot and request that data is inserted at a specific point in time, for example by using a `compare_and_append` directly. If you don't want to fix this in this PR we should open an issue to track it

Reply:
Ill open an issue!

Reply:
broken out here: https://github.com/MaterializeInc/materialize/issues/25349 !